rust-runtime/aws-smithy-runtime-api/src/client/http/request.rs

```diff
@@ -0,0 +1,613 @@
+        HeaderValue::try_from(s.to_string())
+    }
+}
+
+impl TryFrom<http0::HeaderValue> for HeaderValue {
+    type Error = Utf8Error;
+
+    fn try_from(value: http::HeaderValue) -> Result<Self, Self::Error> {
+        HeaderValue::from_http03x(value)
+    }
+}
+
+impl TryFrom<String> for HeaderValue {
+    type Error = HttpError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Ok(HeaderValue::from_http03x(
+            http0::HeaderValue::try_from(value).map_err(HttpError::invalid_header_value)?,
+        )
+        .expect("input was a string"))
+    }
+}
+
+type MaybeStatic = Cow<'static, str>;
+
+fn header_name(name: impl AsHeaderComponent) -> Result<http0::HeaderName, HttpError> {
+    name.repr_as_http03x_header_name().or_else(|name| {
+        name.into_maybe_static().and_then(|cow| match cow {
+            Cow::Borrowed(staticc) => Ok(http0::HeaderName::from_static(staticc)),
+            Cow::Owned(s) => http0::HeaderName::try_from(s).map_err(HttpError::invalid_header_name),
+        })
+    })
+}
+
+fn header_value(value: MaybeStatic) -> Result<HeaderValue, HttpError> {
+    let header = match value {
+        Cow::Borrowed(b) => http0::HeaderValue::from_static(b),
+        Cow::Owned(s) => {
+            http0::HeaderValue::try_from(s).map_err(HttpError::invalid_header_value)?
+        }
+    };
+    header.try_into().map_err(HttpError::new)
+}
+
+#[cfg(test)]
+mod test {
+    use aws_smithy_http::body::SdkBody;
+    use http::header::{AUTHORIZATION, CONTENT_LENGTH};
+    use http::{HeaderValue, Uri};
+
+    #[test]
+    fn headers_can_be_any_string() {
+        let _: HeaderValue = "😹".parse().expect("can be any string");
+        let _: HeaderValue = "abcd".parse().expect("can be any string");
+        let _ = "a\nb"
+            .parse::<HeaderValue>()
+            .expect_err("cannot contain control characters");
+    }
+
+    #[test]
+    fn request_can_be_created() {
+        let req = http::Request::builder()
+            .uri("http://foo.com")
+            .body(SdkBody::from("hello"))
+            .unwrap();
+        let mut req = super::Request::try_from(req).unwrap();
+        req.headers_mut().insert("a", "b");
+        assert_eq!(req.headers().get("a").unwrap(), "b");
+        req.headers_mut().append("a", "c");
+        assert_eq!(req.headers().get("a").unwrap(), "b");
+        let http0 = req.into_http03x().unwrap();
+        assert_eq!(http0.uri(), "http://foo.com");
+    }
+
+    #[test]
+    fn uri_mutations() {
+        let req = http::Request::builder()
+            .uri("http://foo.com")
+            .body(SdkBody::from("hello"))
+            .unwrap();
+        let mut req = super::Request::try_from(req).unwrap();
+        assert_eq!(req.uri(), "http://foo.com/");
+        req.set_uri("http://bar.com").unwrap();
+        assert_eq!(req.uri(), "http://bar.com");
+        let http0 = req.into_http03x().unwrap();
+        assert_eq!(http0.uri(), "http://bar.com");
+    }
+
```

Comment:
Just note somewhere that there are more tests to add. For example, the `Uri::set_endpoint` method.